### ISSUE TITLE
Figure out permissions and gspread

### ISSUE THREAD

**Issue by gpsaggese** [sorrentum/sorrentum#416]
Sonya Nikiforova, [7/13/23 11:34 AM]
Hi GP! There is a bit of a problem with using gspread-pandas for our tasks. 

First of all, it's not in the env anywhere and needs to be installed by each user on their own. But it's not the biggest problem:) 
Second, as I just found out, if one creates a personal Google API project and does not wish to make it public for all to use, then the token will at some point expire, and I can't figure out how to refresh it without completely deleting everything and creating a brand new project. I remember that it was easier when we used this library in Particle because we could create a project from our business account and make it public within the organization, which didn't lead to the token expiration problem.

Sonya Nikiforova, [7/13/23 11:36 AM]
That's not the problem for people who have a crypto-kaizen account, ofc, but I don't and I also assume not everyone who would in theory work using gspread-pandas will have one

GP Saggese, [7/13/23 12:34 PM]
Can we create a public dir for now to keep moving while I investigate?

Sonya Nikiforova, [7/13/23 12:42 PM]
A public dir on a gdrive you mean? But the problem, I think, is different. It is that the Google API application needs to be public. The thing that gives access to the drive and gsheets available to your Google account.

Sonya Nikiforova, [7/13/23 12:44 PM]
If it's a private Google account, and we are making the application that gives access to it public, it's a problem...

**Comment by sonniki** [comment]
Making it a little less dramatic: I figured out you _can_ refresh the token without creating a brand new Google API application/project. For me it was the matter of deleting a folder `~/.config/gspread_pandas/creds`. Then it will ask to authenticate again and you can use your old application after that.

**Comment by yiyunlei** [comment]
Hi, I will try to figure out these gSheet issues this week.
Where is the doc `./docs/gsheet_into_pandas.md`? I can't find it.

**Comment by yiyunlei** [comment]
> Making it a little less dramatic: I figured out you _can_ refresh the token without creating a brand new Google API application/project. For me it was the matter of deleting a folder `~/.config/gspread_pandas/creds`. Then it will ask to authenticate again and you can use your old application after that.

Yeah, that's right. After deleting the old credentials, it should ask to authenticate again. The current issue is to make sure it can authenticate successfully in the Docker environment. I'll do some tests and update the docs.

**Comment by yiyunlei** [comment]
- ETA: None yet.
  - How to authorize Google API inside of Docker.
  - Update docs.

- References:
    - [How to authorize Google API inside of Docker](https://stackoverflow.com/questions/68687358/how-to-authorize-google-api-inside-of-docker)
    - [Set up authentication for Docker](https://cloud.google.com/artifact-registry/docs/docker/authentication#json-key)

**Comment by yiyunlei** [comment]
- ETA: Tomorrow(2023-08-17)
- Keep working on it.

**Comment by yiyunlei** [comment]
> Yeah, that's right. After deleting the old credentials, it should ask to authenticate again. The current issue is to make sure it can authenticate successfully in the Docker environment. I'll do some tests and update the docs.

After spending a lot of time searching for the solution, I finally found a suitable solution from an issue in the repo google-auth-oauthlib:

- https://github.com/googleapis/google-auth-library-python-oauthlib/issues/87

If we want to use Google Oauth in the Docker container (so that we can use Google API), we have to expose one more port to get the access token and refresh token.
- https://github.com/googleapis/google-auth-library-python-oauthlib/issues/87#issuecomment-1559657694

I have tested adding one more port in the jupyter docker, then try to go through the Google Oauth process again, and it works.

@gpsaggese Do you think it's okay to add one more port in the docker? If so, I can create a PR for it.

**Comment by gpsaggese** [comment]
We can certainly open one more port.

Again good job getting down in the weeds and finding a solution. This is how we like to do things.

The only caveat is to understand when to "give up". Sometimes if the solution is too complicated, it might be better to change the problem itself.

In any case, great job.

**Comment by gpsaggese** [comment]
@yiyunlei do the PR and then we can close

**Comment by gpsaggese** [comment]
<img width="991" alt="image" src="https://github.com/sorrentum/sorrentum/assets/33238329/2aad517b-5fc5-4dd1-bc00-c038f0c691ea">

**Comment by yiyunlei** [comment]
I created a PR #566 for it, and the instruction for it is here:
https://docs.google.com/document/d/1dfAVMJZtHtQ3sX7UHwpXNAWPiJs17w0V8FlsOI7a32M/edit?usp=sharing
Then we can pass Google Oauth authentication in the Docker.

We have to use the function I created in the `google_drive_api.py` to do the authentication as I can't change the code in the gspread-pandas repo.

I was trying to change the creds file path for gspread-pandas so that it can use the token file created by the `google_drive_api.py`, but I haven't figured it out and I will try later to see if it's possible.
